keps/sig-apps/20190410-configmap-secret-orchestration.md

```diff
@@ -0,0 +1,214 @@
+KEP editors, SIG Docs, and SIG PM should help to ensure that the tone and content of the `Summary` section is useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Kubernetes added support for higher level orchestration in the form of ReplicaSets/Deployments/DaemonSets/Statefulsets. They implement best practices for things including performing rolling upgrades and ensuring the containers are immutable. The orchestration resources currently do not fully implement a flow that provides immutable infrastructure without extra careful work by the user. Config/Secrets are injected into the immutable containers but aren't tracked along with the roll-forward/roll-back. Since config/secrets can change over time, its often the case where the user updates the config/secret for the new revision of the Deployment but rollbacks fail as they now reference the new config file instead of the one they were deployed with. The other problem with config/secret orchestration is that of orchestrated updates of the immutable infra can only be triggered manually by changes. A config/secret only change can not trigger an upgrade either manually or automatically without making an unnecessarily change to the corresponding orchestration object.
+
+We would like to solve these issues by adding support to the orchestration resources to support orchestration support around config/secrets.
+
+
+## Motivation
+
+Currently there are a mishmash of workarounds to the orchestration objects not supporting orchestration of configmaps/secrets.
+
+Several different solutions or workarounds out of tree are being used. See Alternatives section for details. None of these are widely deployed. Since most workloads need this kind of orchestration, having it as a base Kubernetes functionality would be a significant benefit to users.
+
+### Goals
+
+* Enhance the deployment/statefulset/daemonset with a flag asking for the workload to be automatically updated when a specified configmap/secret changes.
+* Enhance the deployment/statefulset/daemonset with a flag asking for the specified configmap/secret's current state to be snapshotted and follow the orchestrated workload (roll-forward/roll-back)
+* Implementations of the api changes
+
+### Non-Goals
+
+* Out of tree solutions. For applications to be portable, it really needs to be in Kubernetes proper so application developers know the functionality is always there.
+
+## Proposal
+
+The ConfigMapVolumeSource and SecretVolumeSource objects will get two new, optional fields:
+ * snapshot: boolean
+ * watch: boolean
+
+Both will default to false for compatibility.
+
+snapshot can be either true or false. Watch can only be true if snapshot is also true. This ensures immutability.
+
+The fields will be ignored by all objects other then Deployment, DaemonSet and StatefulSet.
+
+DaemonSet/StatefulSet controllers will be modified such that:
```

Comment:
I personally think it might more sense to:
* Have a dedicated controller 
* Make that controller update an annotation on the `podTemplate` based e.G. on configmap/secret resourceVersion and leave the actual rollout to the existing controllers

and as mentioned  before, it will be easier to not create new configmaps/secrets and also raise less scalability/debuggability  concerns

Reply:
So a new controller that also watches Deployments/DaemonSets/Statefulsets and triggers a rolling upgrade via annotation? That could be made to work I think. Wouldn't that put 2x the load on the apiserver though?

yes, it would be easier not to create new configmaps/secrets, but that sidesteps the actual users issue.
Users need to be able to easily update config. That is easiest done in place by editing the existing configmap. Sure you can make everyone upload immutable configmaps, uniquely named, and then use those for the roll forward/backwards. But that requires the user to: 1, keep track of unique names, 2. update the deployment when it changes, 3 manually garbage collect the configmap/secret when its done with.

How does watch work if you force the secret/configmap to be immutable? Theres no reason to watch then I think?

Reply:
> That could be made to work I think. Wouldn't that put 2x the load on the apiserver though?

Why would it put double the load on the apiserver? Caches are shared within the kube-controller-manager.

> How does watch work if you force the secret/configmap to be immutable? Theres no reason to watch then I think?

If you would use this with immutable configmap/secrets it will never do anything, yes

> yes, it would be easier not to create new configmaps/secrets, but that sidesteps the actual users issue.
Users need to be able to easily update config

As mentioned before, yes, having managed, immutable configmaps/secrets that are created whenever a configmap is updated would also be nice but it is an orthogonal issue. I guess if you were to solve the managed configmaps/secrets you wouldn't need the watching anymore. But its a lot more debatable.

Reply:
Agreed the API proposed should be accomodated by its own controller.